### PR TITLE
Python grammar fixes, test samples & improvement offers

### PR DESCRIPTION
**Pull request by inkoit**
Although it seems the PR changes grammar dramatically, in fact, it doesn't. There are only a few minor improvements. All the changes were ported into PT.PM in my local fork, too

- I've changed the order of rules in `PythonParser.g4` for more pleasant reading and testing: top-down in my case. For example, `decorator` could be defined in line 49, but used only in line 172. I've tried to group rules just for better navigation while testing

- added test samples for all the grammar branches coverage. Each file represents a rule or a group of rules

- some fixes:
   - added trailing comma in `typedargslist`
   - fixed typo in word `finally` (was `finaly` -- single `l` character)
   - removed some redundant [parentheses](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L152)
   - [fixed](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L162) `testlist_star_expr` rule so that it always contains comma
   - [fixed](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L172-L173) `assign_part` rule so that you never assign something to `yield` expression
    - [fixed](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L230) `varargslist` rule: permit trailing comma and single `varkwargs`
    - [added](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L240) bare `STAR` to `vardef_parameter`
    - [fixed](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L262) `comparison` rule making `NOT` optional in `NOT IN` operation
    - [fixed](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L299) `dictorsetmaker` to avoid passing comprehension as an argument

- also added some TODO's, because I haven't decided what should I do yet:
    - bare `STAR` parameter [should be followed by named ones only](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L123). For example, the parameter shouldn't be the last one in the list. Shall we make it as a rule or leave it this way? [Here too](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L239)
    - the left part of an augmented assignment should be `test` only, [neither stars nor lists](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L142). I haven't changed it just to avoid significant changes in PT.PM module, because these changes will lead to `expr_stmt` splitting
    - (just thoughts) in [this](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L173) configuration grammar parses the expression `x: int` as the `expr assign_part`, but `: int` is not an assignment
    - i don't really understand why `name` may [be equal](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L202) to `True` or `False`
    - the `call` variant in `trailer` [currently offers](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L202) really wide range of expressions to pass, but [the official documentation](https://docs.python.org/3/reference/expressions.html#calls) doesn't (interpreter, too). Haven't changed due to the same reason
    - some [suggestions](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L370) [to inline](https://github.com/inkoit/grammars-v4/blob/7716b4bf177936c9af9993ffed75daea41fc57f9/python/PythonParser.g4#L381)

Hope it's useful

### CODE REVIEW

**Review by codeFather2 — APPROVED** [review]
LGTM. Thanks for covering the test examples. Just rename the folder from `samples` to `examples`

**Comment by inkoit** [comment]
Done.
And what about TODO's?

File: python/PythonParser.g4 [hunk 1]
```diff
@@ -42,68 +42,102 @@ eval_input
     : testlist NEWLINE*
     ;
 
+stmt
+    : simple_stmt
+    | compound_stmt
+    ;
+
+//---------------- compound statement ----------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:50]: What is the meaning of such comments? I don't think it makes sense here. So, remove it here and below.

File: python/PythonParser.g4 [hunk 2]
```diff
@@ -42,68 +42,102 @@ eval_input
...
 
-def_parameter
-    : named_parameter (ASSIGN test)?
-    | STAR
+// Python 2 : EXCEPT test COMMA name
+// Python 3 : EXCEPT test AS name
+except_clause
+    : EXCEPT (test (COMMA name | AS name)?)? COLON suite
     ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:91]: Remove comment.

File: python/PythonParser.g4 [hunk 3]
```diff
@@ -42,68 +42,102 @@ eval_input
...
 
-named_parameter
-    : name (COLON test)?
+// ------ class definition ---------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:93]: Remove.

File: python/PythonParser.g4 [hunk 4]
```diff
@@ -42,68 +42,102 @@ eval_input
...
+classdef
+    : CLASS name (OPEN_PAREN arglist? CLOSE_PAREN)? COLON suite
     ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:97]: Remove.

File: python/PythonParser.g4 [hunk 5]
```diff
@@ -42,68 +42,102 @@ eval_input
...
 
-//python 2 paramteters
-varargslist
-    : (vardef_parameters COMMA)? varargs (COMMA vardef_parameters)? (COMMA varkwargs)
-    | vardef_parameters
+// ------- function and it's parameters definition ---------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4]: Fix:

```
// function and its parameters definition
```

File: python/PythonParser.g4 [hunk 6]
```diff
@@ -42,68 +42,102 @@ eval_input
...
+// TODO: bare STAR parameter must follow named ones
+def_parameter
+    : named_parameter (ASSIGN test)?
+    | STAR
     ;
 
+named_parameter
+    : name (COLON test)?
+    ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:132]: Remove.

File: python/PythonParser.g4 [hunk 7]
```diff
@@ -42,68 +42,102 @@ eval_input
...
+
+// -------- simple statement -------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:134]: Remove.

File: python/PythonParser.g4 [hunk 8]
```diff
@@ -147,10 +188,17 @@ assign_part
       (yield_expr | testlist)
     ;
 
+exprlist
+    : expr (COMMA expr)* COMMA?
+    ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:194]: Remove.

File: python/PythonParser.g4 [hunk 9]
```diff
@@ -115,21 +149,28 @@ small_stmt
...
+      IMPORT (STAR | OPEN_PAREN import_as_names CLOSE_PAREN | import_as_names)        #from_stmt
     | GLOBAL name (COMMA name)*                                                       #global_stmt
-    | EXEC expr (IN test (COMMA test)?)?                                              #exec_stmt
+    | EXEC expr (IN test (COMMA test)?)?                                              #exec_stmt     // Python 2
     | ASSERT test (COMMA test)?                                                       #assert_stmt
     | NONLOCAL name (COMMA name)*                                                     #nonlocal_stmt // Python 3
     ;
+//----------------------------------------------------------------------------------------------------------------------
 
+// -------- expression statement ---------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:161]: Remove.

File: python/PythonParser.g4 [hunk 10]
```diff
@@ -115,21 +149,28 @@ small_stmt
...
-       IMPORT (STAR | OPEN_PAREN import_as_names CLOSE_PAREN | import_as_names))      #from_stmt
+    | FROM ((DOT | ELLIPSIS)* dotted_name | (DOT | ELLIPSIS)+)
+      IMPORT (STAR | OPEN_PAREN import_as_names CLOSE_PAREN | import_as_names)        #from_stmt
     | GLOBAL name (COMMA name)*                                                       #global_stmt
-    | EXEC expr (IN test (COMMA test)?)?                                              #exec_stmt
+    | EXEC expr (IN test (COMMA test)?)?                                              #exec_stmt     // Python 2
     | ASSERT test (COMMA test)?                                                       #assert_stmt
     | NONLOCAL name (COMMA name)*                                                     #nonlocal_stmt // Python 3
     ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:159]: Remove.

File: python/PythonParser.g4 [hunk 11]
```diff
@@ -147,10 +188,17 @@ assign_part
       (yield_expr | testlist)
     ;
 
+exprlist
+    : expr (COMMA expr)* COMMA?
+    ;
+//----------------------------------------------------------------------------------------------------------------------
+
+// -------- import statement -------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:196]: Remove.

File: python/PythonParser.g4 [hunk 12]
```diff
@@ -162,50 +210,48 @@ dotted_as_names
 dotted_as_name
     : dotted_name (AS name)?
     ;
+//----------------------------------------------------------------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4:213]: Remove.

File: python/PythonParser.g4 [hunk 13]
```diff
@@ -162,50 +210,48 @@ dotted_as_names
...
-    | decorator* (classdef | funcdef)                                              #class_or_func_def_stmt
-    ;
-
-elif_clause
-    : ELIF test COLON suite
-    ;
-
-else_clause
-    : ELSE COLON suite
+// -------------------------- test, lambda and it's parameters ---------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4]:
```antlr
// test, lambda, and its parameters
```

File: python/PythonParser.g4 [hunk 14]
```diff
@@ -162,50 +210,48 @@ dotted_as_names
...
+/*
+ * Warning!
+ * According to https://docs.python.org/3/reference/expressions.html#lambda LAMBDA shouls be followed by
```
Comment (KvanTTT) [inline python/PythonParser.g4]: should

File: python/PythonParser.g4 [hunk 15]
```diff
@@ -276,13 +333,41 @@ integer
...
+// -------------------------- trailer ----------------------------------------------------------------------------------
+// TODO: this way we can pass: `f(x for x in i, a)`, but it's invalid.
+// See: https://docs.python.org/3/reference/expressions.html#calls
 trailer
     : OPEN_PAREN arglist? CLOSE_PAREN
     | OPEN_BRACKET subscriptlist CLOSE_BRACKET
     | DOT name
     ;
 
+// TODO: maybe inline?
```
Comment (KvanTTT) [inline python/PythonParser.g4]: Why?

File: python/PythonParser.g4 [hunk 16]
```diff
@@ -293,46 +378,25 @@ subscript
...
-    : test (comp_for | ASSIGN test)?
-    | (POWER | STAR) test
+comp_iter
+    : comp_for
+    | IF test comp_iter?
     ;
+//----------------------------------------------------------------------------------------------------------------------
 
+// TODO delete it
+// --------------------- !!! WARNING completely unused !!!--------------------------------------------------------------
```
Comment (KvanTTT) [inline python/PythonParser.g4]: So, maybe remove it at all?

**Review by KvanTTT — COMMENTED** [review]
Thank you for the big examples set! But some issues need to be resolved. Mainly, there are useless comments with a big number of hyphens.

I think we can leave current TODOs as it, and resolve them later. Also, I don't know Python very well and cannot resolve them by myself. Maybe somebody else will help us in the future.

File: python/PythonParser.g4 (hunk 1 above)
Comment (inkoit) [reply python/PythonParser.g4:50]: The purpose of the comments is to match a rules subset to a test sample. E.g. `funcdef.py` tests lines 99-132. Found it useful while writing tests (navigating through the grammar). Should I remove them anyway?

File: python/PythonParser.g4 (hunk 14 above)
Comment (inkoit) [reply python/PythonParser.g4]: Fixed

**Review by KvanTTT — APPROVED** [review]
Ok, thanks. I'm going to fix the rest issues by myself.

**Pull request merged**